netplan/libnetplan.py

```diff
@@ -668,6 +672,48 @@ def __next__(self):
         return NetDefinition(self.np_state, next_value)
 
 
+class _NetdefAddressesIterator:
+    _abi_loaded = False
+
+    @classmethod
+    def _load_abi(cls):
+        if cls._abi_loaded:
+            return
+
+        if not hasattr(lib, '_netplan_new_netdef_addresses_iter'):  # pragma: nocover (hard to unit-test against the WRONG lib)
+            raise NetplanException('''
+                The current version of libnetplan does not allow iterating by IP addresses.
+                Please ensure that both the netplan CLI package and its library are up to date.
+            ''')
+        lib._netplan_new_netdef_addresses_iter.argtypes = [_NetplanNetDefinitionP]
+        lib._netplan_new_netdef_addresses_iter.restype = c_void_p
+
+        lib._netplan_netdef_addresses_iter_next.argtypes = [c_void_p]
+        lib._netplan_netdef_addresses_iter_next.restype = c_char_p
+
+        lib._netplan_netdef_addresses_free_iter.argtypes = [c_void_p]
+        lib._netplan_netdef_addresses_free_iter.restype = None
+
+        cls._abi_loaded = True
+
+    def __init__(self, netdef):
+        self._load_abi()
+        self.netdef = netdef
+        self.iterator = lib._netplan_new_netdef_addresses_iter(netdef)
+
+    def __del__(self):
+        lib._netplan_netdef_addresses_free_iter(self.iterator)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        next_value = lib._netplan_netdef_addresses_iter_next(self.iterator)
```

Comment:
thought (non-blocking): I wonder if we should be introducing a new class/type (e.g. `NetplanAddress`), as addresses are not necessarily just strings. They can have a label, lifetime or other options to it, e.g.:

```yaml
network:
  ethernets:
    eth0:
      addresses:
        - "10.0.0.15/24":
            lifetime: 0
            label: "maas"
        - "2001:1::1/64"
```

This is all still considered internal API/Bindings, so we're free to do what we want (for now). If it's not too much effort we could already prepare it to be future-proof by introducing a new class which returns the `netxt_value` string as part of it's `__str__()` method, but also allows having additional properties, like "label" or "lifetime". Those do not need to be implemented, yet. But data passing from libnetplan and storage in the Python class should allow for it.

Reply:
hmm that's something we probably must do now... I just realized that IPs with properties are stored in a separate list... so my current implementation doesn't return them.